lib/streamlit/legacy_caching/caching.py

```diff
@@ -457,30 +499,36 @@ def cache(
     # Support passing the params via function decorator, e.g.
     # @st.cache(persist=True, allow_output_mutation=True)
     if func is None:
-        return lambda f: cache(
-            func=f,
-            persist=persist,
-            allow_output_mutation=allow_output_mutation,
-            show_spinner=show_spinner,
-            suppress_st_warning=suppress_st_warning,
-            hash_funcs=hash_funcs,
-            max_entries=max_entries,
-            ttl=ttl,
-        )
+
+        def wrapper(f: F) -> F:
+            return cache(
+                func=f,
+                persist=persist,
+                allow_output_mutation=allow_output_mutation,
+                show_spinner=show_spinner,
+                suppress_st_warning=suppress_st_warning,
+                hash_funcs=hash_funcs,
+                max_entries=max_entries,
+                ttl=ttl,
+            )
+
+        return wrapper
+    else:
+        non_optional_func = func
```

Comment:
I think it may allow the type-checker to be sure that the type of `non_optional_func` is `F`, whereas it may incorrectly deduce that `func` has type `Optional[F]`. I'll let @areeh confirm, though.

This PR should be good to go once we have this one small nit addressed

Reply:
You're right that I added this to satisfy the type checker. Mypy correctly detects that `func` is no longer `Optional` after the return, but I still got later type errors reported as if it was `Optional` unless I assigned to a new variable (with or without `else`):
```
# Before  return
streamlit/legacy_caching/caching.py:498:17: note: Revealed type is "Union[F`-1, None]"
# After return
streamlit/legacy_caching/caching.py:518:17: note: Revealed type is "F`-1"
# Later uses of func
streamlit/legacy_caching/caching.py:530:20: error: "None" not callable  [misc]
                return func(*args, **kwargs)
```
I'll push the version without this workaround and you can see what your CI has to say about it